presto-parser/src/main/java/com/facebook/presto/sql/parser/ErrorHandler.java

```diff
@@ -0,0 +1,207 @@
+        private final Map<Integer, String> specialRules;
+        private final Map<Integer, String> specialTokens;
+        private final Set<Integer> ignoredRules;
+
+        public Analyzer(TokenStream stream, ATN atn, Vocabulary vocabulary, Map<Integer, String> specialRules, Map<Integer, String> specialTokens, Set<Integer> ignoredRules)
+        {
+            this.stream = stream;
+            this.atn = atn;
+            this.vocabulary = vocabulary;
+            this.specialRules = specialRules;
+            this.specialTokens = specialTokens;
+            this.ignoredRules = ignoredRules;
+        }
+
+        private Set<Integer> process(Map<Integer, Set<String>> candidates, ATNState startState, int tokenIndex, RuleContext next)
+        {
+            int currentRule = startState.ruleIndex;
+
+            // if startState is the beginning or a rule that's in the list of "special" rules, just bail out and
+            // add that as the value
+            if (startState.getStateType() == BLOCK_START || startState.getStateType() == RULE_START) {
+                if (specialRules.containsKey(currentRule)) {
+                    record(candidates, tokenIndex, specialRules.get(currentRule));
+                    return Collections.emptySet();
+                }
+                else if (ignoredRules.contains(currentRule)) {
+                    return Collections.emptySet();
+                }
+            }
+
+            Set<Integer> endTokens = new HashSet<>();
+
+            Queue<ParsingState> activeStates = new ArrayDeque<>();
+
+            activeStates.add(new ParsingState(startState, tokenIndex));
+
+            while (!activeStates.isEmpty()) {
+                ParsingState parsingState = activeStates.poll();
+
+                if (parsingState.state instanceof RuleStopState) {
+                    if (next != null) {
+                        process(candidates, ((RuleTransition) atn.states.get(next.invokingState).transition(0)).followState, parsingState.tokenIndex, next.parent);
+                    }
+                    continue;
+                }
+
+                for (int i = 0; i < parsingState.state.getNumberOfTransitions(); i++) {
+                    Transition transition = parsingState.state.transition(i);
+
+                    if (transition instanceof RuleTransition) {
+                        for (int token : process(candidates, transition.target, parsingState.tokenIndex, null)) {
+                            activeStates.add(new ParsingState(((RuleTransition) transition).followState, token));
+                        }
+                    }
+                    else if (transition.isEpsilon()) {
+                        activeStates.add(new ParsingState(transition.target, parsingState.tokenIndex));
+                    }
+                    else if (transition instanceof WildcardTransition) {
+                        throw new UnsupportedOperationException("not yet implemented: wildcard transition");
```

Comment:
Does this ever happen in out tree?  Regardless, we should have a catch all around all of this code to deal with any bugs like this, so users get at least the old bad messages.

Reply:
No. We don't use wildcard transitions in our grammar.